Classes/common/HTTP/CDTSessionCookieInterceptor.m

```diff
@@ -0,0 +1,147 @@
+//  and limitations under the License.
+//
+
+#import "CDTSessionCookieInterceptor.h"
+#import "CDTLogging.h"
+
+@interface CDTSessionCookieInterceptor ()
+
+@property (nonnull, strong, nonatomic) NSString *cookieRequestBody;
+@property (nonatomic) BOOL shouldMakeCookieRequest;
+@property (nullable, strong, nonatomic) NSString *cookie;
+@property (nonnull, nonatomic, strong) NSURLSession *urlSession;
+
+@end
+
+@implementation CDTSessionCookieInterceptor
+
+- (instancetype)initWithUsername:(NSString *)username password:(NSString *)password
+{
+    self = [super init];
+    if (self) {
+        _cookieRequestBody = [NSString stringWithFormat:@"name=%@&password=%@", username, password];
+        _shouldMakeCookieRequest = YES;
+        NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
+        _urlSession = [NSURLSession sessionWithConfiguration:config];
+    }
+    return self;
+}
+
+- (CDTHTTPInterceptorContext *)interceptRequestInContext:(CDTHTTPInterceptorContext *)context
+{
+    if (self.shouldMakeCookieRequest) {
+        if (!self.cookie) {
+            // get the cookie
+            self.cookie = [self newCookieForRemote:context.request.URL];
+        }
+        [context.request setValue:self.cookie forHTTPHeaderField:@"Cookie"];
+    }
+
+    return context;
+}
+
+- (CDTHTTPInterceptorContext *)interceptResponseInContext:(CDTHTTPInterceptorContext *)context
+{
+    // check the response code.
+    if (context.response.statusCode == 401) {
+        // reset the cookie, let the request interceptor handle getting a new cookie.
+        self.cookie = nil;
+        context.shouldRetry = YES;
+    }
+
+    return context;
+}
+
+- (nullable NSString *)newCookieForRemote:(NSURL *)url
+{
+    NSString *scheme = [url scheme];
+    NSNumber *port = [url port];
+    if (!port) {
+        if ([scheme isEqualToString:@"https"]) {
+            port = @(443);
+        } else {
+            port = @(80);
+        }
+    }
+
+    NSURL *sessionUrl = [NSURL
+        URLWithString:[NSString stringWithFormat:@"%@://%@:%@/_session", scheme, [url host], port]];
+
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:sessionUrl];
+    request.HTTPBody = [self.cookieRequestBody dataUsingEncoding:NSUTF8StringEncoding];
+    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
+
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+    __block NSString *cookie = nil;
+    NSURLSessionDataTask *task = [self.urlSession
+        dataTaskWithRequest:request
+          completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+            // all parameters to block are nullable, annotations removed
+            // to provide compatbility with xcode 6.4 and xcode 7
+            NSHTTPURLResponse *httpResp = (NSHTTPURLResponse *)response;
+            // handle response being nil
```

Comment:
I meant the comment -- it didn't seem to be addressed in the `if` statement, I was wondering whether it was out of date.